src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,527 @@
+             },
+            .data = {
+                 .size = 0,
+             }
+        }
+    };
+
+    tpm2Data = OPENSSL_malloc(sizeof(*tpm2Data));
+    if (tpm2Data == NULL) {
+        ERR(tpm2tss_sym_genkey, TPM2TSS_R_GENERAL_FAILURE);
+        goto error;
+    }
+    memset(tpm2Data, 0, sizeof(*tpm2Data));
+
+    // Set mode, algo, keysize
+    inPublic.publicArea.parameters.symDetail.sym.algorithm = algo;
+    inPublic.publicArea.parameters.symDetail.sym.keyBits.sym = bits;
+    inPublic.publicArea.parameters.symDetail.sym.mode.sym = mode;
+
+    if (password) {
+        DBG("Setting a password for the created key.\n");
+        if (strlen(password) > sizeof(tpm2Data->userauth.buffer) - 1) {
+            goto error;
+        }
+        tpm2Data->userauth.size = strlen(password);
+        memcpy(&tpm2Data->userauth.buffer[0], password,
+               tpm2Data->userauth.size);
+
+        inSensitive.sensitive.userAuth.size = strlen(password);
+        memcpy(&inSensitive.sensitive.userAuth.buffer[0], password,
+               strlen(password));
+    } else
+        tpm2Data->emptyAuth = 1;
+
+    r = init_tpm_parent(&eactx, parentHandle, &parent);
+    ERRchktss(tpm2tss_sym_genkey, r, goto error);
+
+    tpm2Data->parent = parentHandle;
+
+    DBG("Generating the Sym key inside the TPM.\n");
+
+    r = Esys_Create(eactx.ectx, parent,
+                    ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+                    &inSensitive, &inPublic, &allOutsideInfo, &allCreationPCR,
+                    &keyPrivate, &keyPublic, NULL, NULL, NULL);
+    ERRchktss(tpm2tss_sym_genkey, r, goto error);
+
+    DBG("Generated the Sym key inside the TPM.\n");
+
+    tpm2Data->pub = *keyPublic;
+    tpm2Data->priv = *keyPrivate;
+
+    EVP_CIPHER_CTX_set_app_data(cipher, tpm2Data);
+
+    goto end;
+ error:
+    r = -1;
+    if (tpm2Data)
+        OPENSSL_free(tpm2Data);
+
+ end:
+    free(keyPrivate);
+    free(keyPublic);
+
+    if (parent != ESYS_TR_NONE && !parentHandle)
+        Esys_FlushContext(eactx.ectx, parent);
+
+    esys_auxctx_free(&eactx);
+
+    return (r == TSS2_RC_SUCCESS);
+}
+
+static int
+populate_tpm2data(const unsigned char *key, TPM2_DATA **tpm2Data)
+{
+    uint32_t keyHandle = 0;
+
+    /* Use an empty Key */
+    if (key == NULL) {
+        tpm2Data = NULL;
+
+    /* Read persistent key, use for openssl API : EVP_EncryptInit_ex() */
+    } else if (strncmp((char *)key, "0x81", 4) == 0) {
+        sscanf((char *)key, "0x%x", &keyHandle);
+        if (!tpm2tss_tpm2data_readtpm(keyHandle, tpm2Data))
+            return 0;
+
+    /* Read persistent key, use for openssl app : openssl enc */
+    } else if ((key[0] == 0x81) && (key[HANDLE_SIZE/2 + 1] == 0)) {
+        keyHandle = convert_array_hex_to_int(key, HANDLE_SIZE/2);
+        if (!tpm2tss_tpm2data_readtpm(keyHandle, tpm2Data))
+            return 0;
+
+    /* Use blob context */
+    } else {
+        //if (!tpm2tss_tpm2data_read((char *)key, tpm2Data))
```

Comment:
If we want to merge the current version, this should probably be removed.

Reply:
Previous code : 
```c
    /* Use blob context */
    } else {
        //if (!tpm2tss_tpm2data_read((char *)key, tpm2Data))
        //    return 0;
    }
return 1;
```
Current code :
```c
    /* Use blob context */
    } else {
        return 0;
    }
return 1;
```
Because of that, `tpm2_cipher_init_key()` is failing when doing the `test/sserver.sh` with openssl 1.1.0.